pysrc/Table.cpp

```diff
@@ -18,276 +18,120 @@
-            const double* vals = GetNumpyArrayData<double>(f.ptr());
-            Table2D<double,double>::interpolant i = Table2D<double,double>::linear;
-            if (interp == "linear") i = Table2D<double,double>::linear;
-            else if (interp == "floor") i = Table2D<double,double>::floor;
-            else if (interp == "ceil") i = Table2D<double,double>::ceil;
-            else if (interp == "nearest") i = Table2D<double,double>::nearest;
-            else {
-                PyErr_SetString(PyExc_ValueError, "Invalid interpolant");
-                bp::throw_error_already_set();
-            }
-            return new Table2D<double,double>(xargs, yargs, vals, Nx, Ny, i);
-        }
-
-        static void interpMany(const Table2D<double,double>& table2d,
-                               const bp::object& x, const bp::object& y,
-                               const bp::object& vals)
-        {
-            const double* xvec = GetNumpyArrayData<double>(x.ptr());
-            const double* yvec = GetNumpyArrayData<double>(y.ptr());
-            double* valvec = GetNumpyArrayData<double>(vals.ptr());
-            int N = GetNumpyArrayDim(x.ptr(), 0);
-            table2d.interpMany(xvec, yvec, valvec, N);
-        }
-
-        static void interpManyMesh(const Table2D<double,double>& table2d,
-                                   const bp::object& x, const bp::object& y,
-                                   const bp::object& vals)
-        {
-            const double* xvec = GetNumpyArrayData<double>(x.ptr());
-            const double* yvec = GetNumpyArrayData<double>(y.ptr());
-            double* valvec = GetNumpyArrayData<double>(vals.ptr());
-            int Nx = GetNumpyArrayDim(x.ptr(), 0);
-            int Ny = GetNumpyArrayDim(y.ptr(), 0);
-            assert(Nx == GetNumpyArrayDim(vals.ptr(), 0));
-            assert(Ny == GetNumpyArrayDim(vals.ptr(), 1));
-            table2d.interpManyMesh(xvec, yvec, valvec, Nx, Ny);
-        }
+            Table2D::interpolant i = Table2D::linear;
+            if (interp == "floor") i = Table2D::floor;
+            else if (interp == "ceil") i = Table2D::ceil;
+            else if (interp == "nearest") i = Table2D::nearest;
 
-        static bp::tuple Gradient(const Table2D<double,double>& table2d,
-                                   double x, double y)
-        {
-            double dfdx;
-            double dfdy;
-            table2d.gradient(x, y, dfdx, dfdy);
-            return bp::make_tuple(dfdx, dfdy);
+            return new Table2D(x, y, vals, Nx, Ny, i);
         }
 
-        static void GradientMany(const Table2D<double,double>& table2d,
-                                 const bp::object& x, const bp::object& y,
-                                 const bp::object& dfdx, const bp::object& dfdy)
+        static void interpMany(const Table2D& table2d,
+                               size_t ix, size_t iy, size_t ivals, int N)
```

Comment:
Where do these size_t's come from?  Just the type that gets assigned to blah.ctypes.data?  Feels weird, but I guess it works....

Reply:
These are basically `void*` pointers, which isn't something python understands directly, so the trick is to call them `size_t`, and then `reinterpret_cast` them back into the kind of pointer you actually have.  This is what I use in TreeCorr all the time to pass arrays to C++ with `ctypes.data` (which is a `double*`).  I've also already used it in GalSim for the WCS speedup functions `ApplyPV`, `ApplyCD`, etc.

Reply:
Okay, but why `size_t` then?  If it's arbitrary why not just call them `double*` from the start?  (I guess it's not actually arbitrary?)

Reply:
Right.  It's not arbitrary.  `ctypes.data` in python really is returning an integer, since python can't deal with any kind of pointer directly.  So the dance is that numpy's C-layer data object is really a `double*`.  Then in python you can access it as an integer and pass it to your C-layer functions.  Then you recast it back to a `double*`.  I use `size_t`, because that's the C++ name for the integer type that is guaranteed to hold the same number of bytes as a pointer, which makes it safe to do this.

For more information, [here](https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.ndarray.ctypes.html) is the numpy description of the ctypes.data attribute which says it is "A pointer to the memory area of the array as a Python integer."

Reply:
Okay, sounds good.  Thanks for the link.